Toolkit/ArcGISToolkit/PopupController.swift

```diff
@@ -0,0 +1,347 @@
+    ///     If false then you must forward calls from the AGSGeoViewTouchDelegate to the PopupController. Defaults to true.
+    ///   - showAddFeatureButton: If true then a UIBarButtonItem will be added to the navigationItem as a right-hand button.
+    public init(geoViewController: UIViewController, geoView: AGSGeoView, takeOverTouchDelegate: Bool = true, showAddFeatureButton: Bool = true){
+        
+        self.geoViewController = geoViewController
+        self.geoView = geoView
+        //self.addNewFeatureButton = UIBarButtonItem(title: "Add Feature", style: .plain, target: nil, action: nil)
+        self.addNewFeatureButton = UIBarButtonItem(barButtonSystemItem: .add, target: nil, action: nil)
+        
+        super.init()
+        
+        self.addNewFeatureButton.target = self
+        self.addNewFeatureButton.action = #selector(addNewFeatureTap)
+        
+        if showAddFeatureButton{
+            if let items = geoViewController.navigationItem.rightBarButtonItems{
+                geoViewController.navigationItem.rightBarButtonItems = [self.addNewFeatureButton] + items
+            }
+            else{
+                geoViewController.navigationItem.rightBarButtonItem = self.addNewFeatureButton
+            }
+        }
+        
+        if takeOverTouchDelegate{
+            self.geoView.touchDelegate = self
+        }
+        
+        sketchEditor.isVisible = true
+        if let mapView = geoView as? AGSMapView{
+            mapView.sketchEditor = sketchEditor
+        }
+    }
+    
+    private var addingNewFeature : Bool = false
+    
+    @objc private func addNewFeatureTap(){
+        
+        // if old pvc is being shown still for some reason, dismiss it
+        self.cleanupLastPopupsViewController()
+        
+        guard let map = (geoView as? AGSMapView)?.map else{
+            return
+        }
+        
+        let templatePicker = TemplatePickerViewController(map: map)
+        templatePicker.delegate = self
+        
+        let navigationController = UINavigationController(rootViewController: templatePicker)
+        navigationController.modalPresentationStyle = .formSheet
+        UIApplication.shared.topViewController()?.present(navigationController, animated: true)
+    }
+    
+    
+    private func cleanupLastPopupsViewController(){
+        unselectLastSelectedFeature()
+        
+        // if old pvc is being shown still for some reason, dismiss it
+        if pvc?.view?.window != nil {
+            if pvc == geoViewController?.navigationController?.topViewController{
+                _ = geoViewController?.navigationController?.popViewController(animated: true)
+            }
+            else if pvc == geoViewController?.presentedViewController{
+                pvc?.dismiss(animated: true, completion: nil)
+            }
+        }
+        
+        // cleanup last time
+        lastPopupQueries?.forEach{ $0.cancel() }
+        pvc = nil;
+        lastPopupQueries = [AGSCancelable]()
```

Comment:
why? I thought creating a new array would be faster?

Reply:
I haven't benchmarked it or anything, but consider what needs to happen. In both cases all of the elements need to be discarded. Also in both cases, some memory needs to be freed. In the case of calling `removeAll()`, the array likely resizes the currently allocated memory to the default size allocated by a new array, but no new memory is allocated. In the case of assigning a new array, all of the existing memory is freed and new memory must be allocated. There's also some release and retain calls in there because `Array` internally uses a reference type for its storage (which is what makes copy on write possible for a value type). So I'm fairly certain that removing all elements is faster than replacing the current instance with a new one.

Reply:
cool